### PR TITLE
Added graph_difference method to SnowShuGraph class to handle increme…

### PR DESCRIPTION
**Pull request by Aesthet**
…ntal replica builds. Added needed flag for incremental builds into the main module. Updated with eq/hash functions classes like Relation and Materialization. Updated _safe_query method of BaseSQLAdapter to accept database as an optional parameter to use as connection's database_override. Updated mock imports, because mock is now part of the Python standard library, available as unittest.mock in Python 3.3 onwards.

### CODE REVIEW

File: snowshu/core/main.py [hunk 1]
```diff
@@ -82,16 +82,23 @@ def source(filename):
               default=None,
               help="Overrides the replica name found in replica.yml")
 @click.option(
-    '--barf',
+    '--barf', '-b',
     is_flag=True,
     help="outputs the source query sql to a local folder snowshu_barf_output")
+@click.option(
```
Comment (eculicny) [inline snowshu/core/main.py:88]: Probably should save the parameters implementation for the subsequent ticket as discussed

File: snowshu/core/graph.py [hunk 2]
```diff
@@ -1,28 +1,89 @@
...
+
+        """
+        if isinstance(source_catalog, (tuple, list, set)):
+            source = list(source_catalog)
+        elif isinstance(source_catalog, SnowShuGraph):
+            source = list(source_catalog.graph.nodes)
+        else:
+            source = list(source_catalog.nodes)
+        non_isolated_relations = []
+        for relation in configs.specified_relations:  # noqa pylint: disable=too-many-nested-blocks
```
Comment (eculicny) [inline snowshu/core/graph.py:50]: At the expense of comparing raw catalogs, we should use the graph's edges to determine the connections between nodes.

**Review by eculicny — DISMISSED** [review]
As discussed, we can save the command options for follow up work, and simplify the connected nodes check with the existing graph.

If we ever need to compare the raw source and target catalogs while taking expected connections into account, we should probably refactor the logic in the `build_graph` function.

File: Dockerfile [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ## docker image is intended for development environments only.
 ## to use snowshu in production install the package via pip
-FROM python:3.7-slim-buster
+FROM python:3.8-slim-buster
```
Comment (Aesthet) [inline Dockerfile:3]: This is not needed as I removed usage of walrus operators - should I revert this change?

Comment (eculicny) [reply Dockerfile:3]: Let's revert it for now since the `setup.py` seems to only expect python version 3.7. We can work on upgrading/updating that dependency later.

**Review by eculicny — COMMENTED** [review]
Looks good! Need to wait until we can verify that the integration tests are passing before approving.

File: Dockerfile (hunk 3 above)
Comment (Aesthet) [reply Dockerfile]: done

**Review by jedibrillo — APPROVED** [review]

**Review by fleischer89 — APPROVED** [review]
